### DISCUSSION TITLE
How to support react-refresh for multiple react dev servers on one page in micro-frontend setup

### DISCUSSION THREAD

**Discussion by goatfryed** [vitejs/vite-plugin-react#185] [Q&A]
I'm currently working on a micro frontend playground where I'm trying out vite + react + micro frontends.

I'm just starting and I have a container application build with react and a single app. The app exposes a web component build with react that I want to include.

The project can be found [here](https://github.com/goatfryed/microfrontend-playground/tree/issues/hmr-dev-setup)

The container's index.html includes the app-one entry point served though its dev server. Since both the container and the app-one use react, I get the error that the @react-refresh runtime is included twice.

Is there a way to make this setup work with HMR? I have researched how to externalize the @react-refresh module so that it's included only once, but I haven't found a way yet. I found that for backend integration setups, sometimes the base must be configured, but this would affect all code, not just  @react-refresh. Likewise, It would be at least better to disable @react-refresh and have HMR for all things except component refresh, but that also doesn't seem possible atm.

Is there a way currently to make this work? Should this be a feature request instead?

**Comment by ArnaudBarre** [comment]
Is there a reason why you start multiple server and don't import the code from other apps via JS imports?

Related answer: https://github.com/vitejs/vite-plugin-react-swc/issues/87#issuecomment-1518675123

**Comment by goatfryed** [comment]
Yes, the reason is that I'm working on a micro frontend setup that should be work with runtime instead of build time integration.
To include them via js import would go against the architecture pattern I'm trying to apply.

I guess this issue would also effect anyone trying to build single-spa react apps with vite. I'm trying to achieve something similar, I'm setting up some things for myself for educational purpose

Thanks for the link! didn't missed that one, but yes it's exactly my issue. I'm trying the proposed solution there

**Comment by Relaxe111** [comment]
We have a similar approach, we are using it in production for over one year, hmr is working ok. We are using qiankun with some modification of the vite-plugin-qiankun. Prior to v 4 it worked with no problem, now we are getting an `throw` error because one restricted of the react-refresh only once per window, while in our project window is shared in browser among all mounted apps.

**Comment by ArnaudBarre** [comment]
I released 4.0.3 without this check!
